openlibrary/plugins/upstream/addtag.py

```diff
@@ -0,0 +1,228 @@
+
+logger = logging.getLogger("openlibrary.tag")
+
+
+class addtag(delegate.page):
+    path = '/tag/add'
+
+    def GET(self):
+        """Main user interface for adding a tag to Open Library."""
+
+        if not self.has_permission():
+            raise common.PermissionDenied(message='Permission denied to add tags')
+
+        return render_template('tag/add', recaptcha=get_recaptcha())
+
+    def has_permission(self) -> bool:
+        """
+        Can a tag be added?
+        """
+        # return web.ctx.user and (
+        #     web.ctx.user.is_usergroup_member('/usergroup/super-librarians')
+        # )
+        return web.ctx.site.can_write("/tag/add")
+
+    def POST(self):
+        i = web.input(
+            tag_name="",
+            tag_type="",
+            tag_description="",
+            tag_plugins="",
+        )
+
+        if spamcheck.is_spam(i, allow_privileged_edits=True):
+            return render_template(
+                "message.html", "Oops", 'Something went wrong. Please try again later.'
+            )
+
+        if not web.ctx.site.get_user():
+            recap = get_recaptcha()
+            if recap and not recap.validate():
+                return render_template(
+                    'message.html',
+                    'Recaptcha solution was incorrect',
+                    'Please <a href="javascript:history.back()">go back</a> and try again.',
+                )
+
+        i = utils.unflatten(i)
+        match = self.find_match(i)  # returns None or Tag (if match found)
+
+        if len(match) > 0:
+            # tag match
+            return self.tag_match(match)
+        else:
+            # no match
+            return self.no_match(i)
+
+    def find_match(self, i: web.utils.Storage):
+        """
+        Tries to find an existing tag that matches the data provided by the user.
+        """
+
+        q = {'type': '/type/tag', 'name': i.tag_name, 'tag_type': i.tag_type}
+        match = list(web.ctx.site.things(q))
+        return match
+
+    def tag_match(self, match: list) -> NoReturn:
+        """
+        Action for when an existing tag has been found.
+        Redirect user to the found tag's edit page to add any missing details.
+        """
+        tag = web.ctx.site.get(match[0])
+        raise safe_seeother(tag.key + "/edit")
+
+    def no_match(self, i: web.utils.Storage) -> NoReturn:
+        """
+        Action to take when no tags are found.
+        Creates a new Tag.
+        Redirects the user to the tag's home page
+        """
+        key = web.ctx.site.new_key('/type/tag')
+        web.ctx.path = key
+        web.ctx.site.save(
+            {
+                'key': key,
+                'name': i.tag_name,
+                'tag_description': i.tag_description,
+                'tag_type': i.tag_type,
+                'tag_plugins': json.loads(i.tag_plugins or "[]"),
+                'type': dict(key='/type/tag'),
+            },
+            comment='New Tag',
+        )
```

Comment:
It would be nice if this was it's own function called `create_tag` which returns a key and then we just redirect to the key

Reply:
We might want to add this functionality as a classmethod to the Tag upstream model itself so we can call `Tag.create(...)`